File: rsdupdater/docker.py
import subprocess
from pathlib import Path


def login(user, password):
    try:
        subprocess.run(args=f"docker login -u={user} --password-stdin".split(' '), input=password.encode('utf-8'))
    except subprocess.CalledProcessError:
        print("docker login failed")
        return False
    return True


def logout():
    try:
        subprocess.run(args="docker logout".split(' '))
    except subprocess.CalledProcessError:
        print("docker logout failed")
        return False
    return True


def run_cmd(cmd):
    print(f"run: {' '.join(cmd)}")
    try:
        process = subprocess.run(args=cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, check=True)
    except subprocess.CalledProcessError as ex:
        print("FAILED")
        print(ex.stdout.decode('utf-8'))
        return False
    print(process.stdout.decode("utf-8"))
    return True


def build(work_dir: Path, dockerfile: str, branch: str, re_commit: str, arch: str):
    repo = "iceflower/redeclipse-server"

    success = run_cmd(
        ["docker", "build", "--build-arg", f"BRANCH={branch}", "--build-arg", f"RECOMMIT={re_commit}", "-t",
         f"{repo}:{arch}-{branch}", "-f", f"{dockerfile}", f"{str(work_dir)}"])
    if not success:
        return False
    success = run_cmd(["docker", "push", f"{repo}:{arch}-{branch}"])
    if not success:
        return False
    success = run_cmd(
        ["docker", "manifest", "create", f"{repo}:{branch}", f"{repo}:amd64-{branch}", f"{repo}:arm64-{branch}"])
    if not success:
        print("Manifest create failed.")
        #return False
    success = run_cmd(["docker", "manifest", "annotate", f"{repo}:{branch}", f"{repo}:arm64-{branch}", "--variant", "v8"])
    if not success:
        print("Manifest annotate failed.")
        # return False
    success = run_cmd(["docker",  "manifest",  "push", "--purge", f"{repo}:{branch}"])
    if not success:
        print("Manifest push failed.")
        # return False
    return True

File: go-github-webhook/pkg/updater/docker.go
package updater

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

func runCmd(name string, arg []string, stdin *string) bool {
	fmt.Println("RUN: " + name + " " + strings.Join(arg, " "))
	cmd := exec.Command(name, arg...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if stdin != nil {
		cmd.Stdin = strings.NewReader(*stdin)
	}

	if err := cmd.Run(); err != nil {
		fmt.Println("FAILED: " + name + " " + strings.Join(arg, " "))
		fmt.Println(err.Error())
		return false
	}
	return true
}

func dockerLogin(user string, password string) bool {
	return runCmd("docker", []string{"login", "--username", user, "--password-stdin"}, &password)
}

func dockerLogout() bool {
	return runCmd("docker", []string{"logout"}, nil)
}

func dockerBuild(workDir string, repo string, dockerfile string, ref string, arch string, reCommit string) bool {
	dockerTag := ref[strings.LastIndex(ref, "/")+1:]
	dockerName := repo + ":" + dockerTag
	dockerArchName := repo + ":" + arch + "-" + dockerTag


	success := runCmd("docker", []string{"build", "--build-arg", "TAG=" + dockerTag, "--build-arg", "RECOMMIT=" + reCommit, "-t", dockerArchName, "-f", dockerfile, workDir}, nil)
	if !success {
		return false
	}
	success = runCmd("docker", []string{"push", dockerArchName}, nil)
	if !success {
		return false
	}
	runCmd("docker", []string{"manifest", "create", dockerName, repo + ":amd64-" + dockerTag, repo + ":arm64-" + dockerTag}, nil)
	runCmd("docker", []string{"manifest", "annotate", dockerName, repo + ":arm64-" + dockerTag, "--variant", "v8"}, nil)
	runCmd("docker", []string{"manifest", "push", "--purge", dockerName}, nil)
	return true
}
